substratevm/src/com.oracle.svm.hosted/src/com/oracle/svm/hosted/image/NativeBootImage.java

```diff
@@ -952,4 +973,299 @@ protected NativeTextSectionImpl(RelocatableBuffer relocatableBuffer, ObjectFile
         protected final ObjectFile objectFile;
         protected final NativeImageCodeCache codeCache;
     }
+
+    /**
+     * implementation of the DebugInfoProvider API interface
+     * that allows type, code and heap data info to be passed to
+     * an ObjectFile when generation of debug info is enabled.
+     */
+    private class NativeImageDebugInfoProvider implements DebugInfoProvider {
+        private final NativeImageCodeCache codeCache;
+        private final NativeImageHeap heap;
+        private final Iterator<Map.Entry<HostedMethod, CompilationResult>> codeCacheIterator;
+        private final Iterator<Map.Entry<Object, ObjectInfo>> heapIterator;
+
+        NativeImageDebugInfoProvider(NativeImageCodeCache codeCache, NativeImageHeap heap) {
+            super();
+            this.codeCache = codeCache;
+            this.heap = heap;
+            this.codeCacheIterator = codeCache.compilations.entrySet().iterator();
+            this.heapIterator = heap.objects.entrySet().iterator();
+        }
+
+        @Override
+        public DebugTypeInfoProvider typeInfoProvider() {
+            return () -> new Iterator<DebugTypeInfo>() {
+                @Override
+                public boolean hasNext() {
+                    return false;
+                }
+
+                @Override
+                public DebugTypeInfo next() {
+                    return null;
+                }
+            };
+        }
+
+        @Override
+        public DebugCodeInfoProvider codeInfoProvider() {
```

Comment:
How about something more modern than `Iterable`. E.g. like:
```
@Override
public Stream<DebugCodeInfo> codeInfoProvider() {
    return compilations.entrySet().stream().map(NativeImageDebugCodeInfo::new);
}
```

Reply:
Yes, that would be neater.

Reply:
@olpaw I just pushed another fix where I modified the DebugInfo interfaces to use Stream and tweaked the producers and consumers accordingly.